dds/DCPS/DataWriterImpl.cpp

```diff
@@ -2830,6 +2822,419 @@ void DataWriterImpl::transport_discovery_change()
+  InstanceValuesToHandles::const_iterator pos = instance_values_to_handles_.find(dummy_rch);
+  dummy_rch._retn();
+  return pos;
+}
+
+DDS::ReturnCode_t DataWriterImpl::get_or_create_instance_handle(
+  DDS::InstanceHandle_t& handle,
+  const Sample& sample,
+  const DDS::Time_t& source_timestamp)
+{
+  ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex, guard, get_lock(), DDS::RETCODE_ERROR);
+
+  handle = lookup_instance(sample);
+  if (handle == DDS::HANDLE_NIL || !get_handle_instance(handle)) {
+    Sample_rch copy = sample.copy(Sample::ReadOnly, Sample::KeyOnly);
+#if defined(OPENDDS_SECURITY) && OPENDDS_HAS_DYNAMIC_DATA_ADAPTER
+    DDS::DynamicData_var dynamic_data = copy->get_dynamic_data(dynamic_type_);
+    DDS::Security::SecurityException ex;
+    if (dynamic_data && security_config_ &&
+        participant_permissions_handle_ != DDS::HANDLE_NIL &&
+        !security_config_->get_access_control()->check_local_datawriter_register_instance(participant_permissions_handle_, this, dynamic_data, ex)) {
+      if (log_level >= LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE,
+                   "(%P|%t) NOTICE: DataWriterImpl::get_or_create_instance_handle: unable to register instance SecurityException[%d.%d]: %C\n",
+                   ex.code, ex.minor_code, ex.message.in()));
+      }
+      return DDS::Security::RETCODE_NOT_ALLOWED_BY_SECURITY;
+    }
+#endif
+
+    // don't use fast allocator for registration.
+    const TypeSupportImpl* const ts = get_type_support();
+    Message_Block_Ptr serialized(serialize_sample(*copy));
+    if (!serialized) {
+      if (log_level >= LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: %CDataWriterImpl::get_or_create_instance_handle: "
+          "failed to serialize sample\n", ts->name()));
+      }
+      return DDS::RETCODE_ERROR;
+    }
+
+    // tell DataWriterLocal and Publisher about the instance.
+    const DDS::ReturnCode_t ret = register_instance_i(handle, move(serialized), source_timestamp);
+    // note: the WriteDataContainer/PublicationInstance maintains ownership
+    // of the marshalled sample.
+    if (ret != DDS::RETCODE_OK) {
+      handle = DDS::HANDLE_NIL;
+      return ret;
+    }
+
+    if (!insert_instance(handle, copy)) {
+      handle = DDS::HANDLE_NIL;
+      if (log_level >= LogLevel::Notice) {
+        ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: %CDataWriterImpl::get_or_create_instance_handle: "
+           "insert instance failed\n", ts->name()));
+      }
+      return DDS::RETCODE_ERROR;
+    }
+
+    send_all_to_flush_control(guard);
+  }
+
+  return DDS::RETCODE_OK;
+}
+
+DDS::ReturnCode_t DataWriterImpl::instance_must_exist(
+  const char* const method_name,
+  const Sample& sample,
+  DDS::InstanceHandle_t& instance_handle,
+  bool remove)
+{
+  OPENDDS_ASSERT(sample.key_only());
+
+  ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex, guard, get_lock(), DDS::RETCODE_ERROR);
+
+  const InstanceValuesToHandles::const_iterator pos = find_instance(sample);
```

Comment:
```suggestion
  const InstanceValuesToHandles::iterator pos = find_instance(sample);
```

This should fix the CI errors.  It's a quirk of the pre-C++11 standard library that `map::erase` doesn't take a `const_iterator`.

Reply:
Yea, I forgot about that. `erase` is a lot better in `C++11` between that and returning the next iterator.